docs/lifecycle-management.md

```diff
@@ -0,0 +1,185 @@
+3. `GET` $c1/stop
+   → `DELETE` $i1
+4. `GET` $c2/start
+   → creates instance `POST` $i2
+   → updates (not .state) `PATCH` $c2 from $i2
+5. `GET` $c2/snapshot
+   → creates `POST` $c3 (without autoSave) from $i2
+6. `GET` $c2/save
+   → updates (not state) `PATCH` $c2 from $i2
+7. `GET` $c2/stop
+   → DELETE $i2
+8. `GET` $c2/start
+   → creates instance POST $i3
+   → updates (not .state) `PATCH` $c2 from $i3
+9. `PATCH` $i3
+   → hot-adds device to domain
+   → updates (not .state) `PATCH` $c2 from $i3
+10. `PATCH` $c3
+
+
+### Observations
+* Expansion is performed at runtime
+* Imperative on config, declarative on instance
+* Hot-plugging by modifying the instance
+* The dynamic instance data can be reflected in the VM.state.domain fields
+
+
+## ICID-E - Extended Option “Imperative-Config-Instance-Dance”
+Like Option “Imperative-Config-Instance-Dance” - but with an added initial
+expansion
+
+### Observations
+* Potentially domain modification can be done without runtime
+   * Assumption that the component used for expanding and checking the
+     out-of-runtime spec does comply with the runtime component to run the spec
+     IOW assumption is that libvirt in expansion creates valid expanded domain
+
+
+## Notes
+
+### Presets and flavors
+Presets look unsuited for flavors. The requirements of flavors are to be able
+to set certain defaults for specific hardware, i.e. use specific busses (i.e.
+virtio) or device models (i.e. e1000) by default. But presets only allow to add
+specific branches to the tree (if you look at the specification as a tree).
+Presets do not allow to user certain values as defaults for arbitrary leaves
+in the tree (which is the defaulting to certain values in certain places).
+
+### UI friendly spec modification
+For a UI it would be nice if spec changes (creating, updating [hot-plugging])
+could be done without runtime interactions because of performance reasons.
+
+
+## Discussion
+
+### Overview
+In general the presented approaches could meet the requirements, an overview:
+
+| Requirement        | ICID | ICID-E |
+| Non running VMs    | Config | |
+| Create then launch | Config and instance | |
+| Flavor             | Special crafted Config | |
+| Expanding          | Performed by runtime | |
+| Detailed           | Detailed Config | |
+| Dynamic data       | .state field of instance | |
+| Defaulting         | (Needs improvement) | |
+| Dry-model-for-UI   | - | + |
+	
+
+### Defaulting - Yes, but how to do it better
+The defaulting is currently an implicit part of the runtime.
+To be able to have saner defaulting, this should probably be externalized.
+
+This also relates to the expansion and validation processes.
+
+Questions are
+   * Should this be a property of the runtime?
+   * Can this be handled on an API level?
+
+
+### ICID Details
+
+Pro
+   * Pretty much all requirements can be met
+
+Cons
+   * A lot of dance between config and instance
```

Comment:
I think we can make it very easy when considering the use-case. One is: I care about reproducibe VM Configs, the other is: I don't care.

A higher level managment plane could either make that decision either per VM, or per application type. In both cases, you can always post your sparse VM Spec or sparse VM Spec updates. 

In one case you get back the fully populated spec, and by default only see that one. So when you later on do a hotplug, you fetch the VM, get the full spec, add your sparse device description and post it again. The management will expand for you the device and give you the fully expanded spec back. Internally different representations or revisions can be used to keep track of changes. 

You could still maintain a complete sparse VM description, if requried, and only face it to the user, if explicitly requests. There might be cases, where you don't have the disks anymore and want your sparse config back, but I don't really see that use-case. In the full-VM-spec scenaro I would expect that you are more interested in an easy way of adding or removing things without thinking about the details, but not about getting back the sparse version.

In the other case you always post the sparse VM and get back the sparse VM. That's basically it. You an still use common code under the hood for revision tracking, ...